src/services/huaweicloud/obs.rs

```diff
@@ -0,0 +1,324 @@
+    write!(&mut s, "{}", canonicalize_resource(req, cred)?)?;
+
+    debug!("string to sign: {}", &s);
+
+    Ok(s)
+}
+
+/// ## Reference
+///
+/// - [Authentication of Signature in a Header](https://support.huaweicloud.com/intl/en-us/api-obs/obs_04_0010.html)
+fn canonicalize_header(req: &impl SignableRequest) -> Result<String> {
+    let mut headers = req
+        .headers()
+        .iter()
+        // Filter all header that starts with "x-obs-"
+        .filter(|(k, _)| k.as_str().starts_with("x-obs-"))
+        // Convert all header name to lowercase
+        .map(|(k, v)| {
+            (
+                k.as_str().to_lowercase(),
+                v.to_str().expect("must be valid header").to_string(),
+            )
+        })
+        .collect::<Vec<(String, String)>>();
+
+    // Sort via header name.
+    headers.sort_by(|x, y| x.0.cmp(&y.0));
+
+    Ok(headers
+        .iter()
+        // Format into "name:value"
+        .map(|(k, v)| format!("{}:{}", k, v))
+        .collect::<Vec<String>>()
+        // Join via "\n"
+        .join("\n"))
+}
+
+/// ## Reference
+///
+/// - [Authentication of Signature in a Header](https://support.huaweicloud.com/intl/en-us/api-obs/obs_04_0010.html)
+fn canonicalize_resource(req: &impl SignableRequest, _cred: &Credential) -> Result<String> {
+    let bucket_name = parse_bucket_name(req.host());
+
+    let mut params: Vec<(Cow<'_, str>, Cow<'_, str>)> =
+        form_urlencoded::parse(req.query().unwrap_or_default().as_bytes()).collect();
+    // Sort by param name
+    params.sort();
+
+    let mut s = String::new();
+
+    write!(&mut s, "/{}", bucket_name.to_string())?;
+    write!(&mut s, "{}", req.path())?;
+
+    let params_str = params
+        .iter()
+        .map(|(k, v)| format!("{}={}", k, v))
+        .collect::<Vec<String>>()
+        .join("&");
+
+    if !params_str.is_empty() {
+        write!(s, "?{}", params_str)?;
+    }
+
+    Ok(s.to_string())
+}
+
+enum ParseBucketNameResult {
+    CustomBucketDomain(String),
+    DefaultDomainWithBucket(String),
+    DefaultDomainWithoutBucket,
+}
+
+impl ToString for ParseBucketNameResult {
+    fn to_string(&self) -> String {
+        match self {
+            ParseBucketNameResult::CustomBucketDomain(d) => d.clone(),
+            ParseBucketNameResult::DefaultDomainWithBucket(d) => d.clone(),
+            ParseBucketNameResult::DefaultDomainWithoutBucket => "".to_string(),
+        }
+    }
+}
+
+lazy_static! {
+    static ref RE: Regex = Regex::new(r"(([\w\W]+).)?obs.[\w\W]+.myhuaweicloud.com").unwrap();
+}
+
+fn parse_bucket_name(host: &str) -> ParseBucketNameResult {
+    if let Some(captures) = RE.captures(host) {
+        if let Some(bucket) = captures.get(2) {
+            ParseBucketNameResult::DefaultDomainWithBucket(bucket.as_str().to_string())
+        } else {
+            ParseBucketNameResult::DefaultDomainWithoutBucket
+        }
+    } else {
+        ParseBucketNameResult::CustomBucketDomain(host.to_string())
+    }
+}
```

Comment:
Please implement unit tests based on examples provided by obs docs: https://obs-community.obs.cn-north-1.myhuaweicloud.com/sign/en-us/header_signature.html

Reply:
Add a unit test for Signer.

Reply:
Add a unit test for Signer.